### ISSUE TITLE
Empty rewards list causes no episode was completed since last summary.

### ISSUE THREAD

**Issue by alessiabertugli** [Unity-Technologies/ml-agents#1169]
Hi, after a lot of training step without any issue (the cumulative rewards increases, the model trained with PPO learns) it happens that the message "no episode was completed since last summary" arises.

I read the code and this message is caused when the list of the received rewards is empty.

However, I'm logging the calls to `AddReward` and I do add rewards. It looks like "something" happened to the model and it get stuck.

I say the model because even when changing the environment a little bit, nothing changes.
Instead, if I restart the train with the same model and the same environment everything starts to work again. But I'm not sure that this situation can happen again (hence, it trains for a lot of steps, but ti could die letting the model without any reward).

Any suggestion on why this happens? Thanks

**Comment by xiaomaogy** [comment]
One possible cause is that although it does add reward, no episode actually has ended. You can try setting a number for your Academy's max steps, so that you know the episode will end after certain steps.

**Comment by alessiabertugli** [comment]
I tried different combinations of agent max steps and academy max steps, but the point is that I don't understand which functions I have to call to reset the game.
I want reset all the game (my agent and also the environment) when the agent wins (touching objects in the correct way) or loses (touching an obstacle). In both this cases I want the end of the episode.
I tried different ways but I'm not sure which is the correct one.

**Comment by vincentpierre** [comment]
@alessiabertugli Even though you use AddReward in your code, the summaries are only printed when an episode is terminated. At that point, the cumulative reward is reported. If no agent has terminated an episode (wins or looses) since the last summary, them the message `no episode was completed since last summary` will appear. I would recommend increasing the `summary-freq` in the trainer config file in order to make sure enough episodes have been completed before printing the summary. If the message `no episode was completed since last summary` appears continuously, it can mean that your agents are stuck and do not terminate episodes anymore (this is where setting max step to some non-zero value can be useful). To answer your second question, you just need to set the agent to done and the agent will automatically reset at the next step.
I hope this helps.

**Comment by alessiabertugli** [comment]
I tied to set agent max steps and academy max steps in different ways: both to 0, both different from 0 (1000) and one to 0 and the other > 0. I want to call AgentReset() in Rewards functions because I want to reset agent position and environment when my agent wins or looses (and not reset only the cumulative reward) and AcademyReset(), but I tried also to call only Done() like examples projects. All these tests seem to work and the cumulative reward increases quickly since "no episode was completed since last summary" appears continuously.
Only two tests were successful with ml-agents v0.4b, agent max steps = 1000 and academy max steps = 1000, but I tried the same configuration with ml-agents v0.5 and it gives me the same problem.
Now I'm using Unity 2018.2.6f1 on Arch linux with ml-agents v0.5.
This is my code:

 public override void AgentAction(float[] vectorAction, string textAction)
    {   
       //Find actual linear distance between target and endEffector
        dirToTarget = target.transform.position - tooltip.transform.position;
    
        MoveAgent(vectorAction);
        RewardFunctionMovingTowards(); //For direction
        RewardFunctionFacingTarget(); //For orientation
        RewardFunctionPositionAndOrientation(target);
        RewardTouchedObject();
        RewardTouchedSupport();
    }
    /// <summary>
    /// Reward touching target.
    /// </summary>
    void RewardTouchedObject(){
        if(suctionHead.GetComponent<Contact>().touching){
            Monitor.Log("Goal reached", "Yeah");
            Debug.Log("Touched "+ DateTime.Now.ToString("h:mm:ss.fff tt"));
            AddReward(1f);
            //AgentReset();
            Done();
        }
    }
    /// <summary>
    /// Reward touching support.
    /// </summary>
    void RewardTouchedSupport(){
         if(suctionHead.GetComponent<Contact>().touchingSupport || surfaceEndEff.GetComponent<Contact>().touchingSupport){
            AddReward(-1f);
            Debug.Log("FAIL "+ DateTime.Now.ToString("h:mm:ss.fff tt"));
            //AgentReset(); 
            Done();
        }
    }
    /// <summary>
    /// Reward moving towards target & Penalize moving away from target.
    /// </summary>
    void RewardFunctionMovingTowards()
    {
        movingTowardsDot = Vector3.Dot(ObjVelocity, dirToTarget.normalized);
        AddReward(0.03f * movingTowardsDot);
        Monitor.Log("Reward moving towards target", 0.03f * movingTowardsDot);
    }
    /// <summary>
    /// Reward facing target & Penalize facing away from target.
    /// </summary>
    void RewardFunctionFacingTarget()
    {   
        facingDot = Vector3.Dot(dirToTarget.normalized, tooltip.transform.forward);
        AddReward(-0.01f * facingDot);
        Monitor.Log("Reward facing target", -0.01f * facingDot);
    }
    /// <summary>
    /// Positive reward if agent takes block in the correct way
    /// </summary>
    
    //Try to give rewards using actual position and orientation of tooltip.
    void RewardFunctionPositionAndOrientation(GameObject blockToGrasp) {
    
        if(tooltip.transform.position.x < (blockToGrasp.transform.position.x + 0.1f)
        && tooltip.transform.position.x > (blockToGrasp.transform.position.x - 0.1f)
        && tooltip.transform.position.y < (blockToGrasp.transform.position.y + 0.1f)
        && tooltip.transform.position.y > (blockToGrasp.transform.position.y - 0.1f)
        && tooltip.transform.position.z > (blockToGrasp.transform.position.z + 0.01)
        && tooltip.transform.position.z < (blockToGrasp.transform.position.z + 0.02) 
        && transform.rotation.eulerAngles.y < (blockToGrasp.transform.rotation.eulerAngles.y + 2.0f)
        && transform.rotation.eulerAngles.y > (blockToGrasp.transform.rotation.eulerAngles.y - 2.0f)
        && transform.rotation.eulerAngles.z < (90.0f + 2.0f)
        && transform.rotation.eulerAngles.z > (90.0f - 2.0f)){
            Debug.Log("WIN");
            AddReward(10f);
            //AgentReset();
            Done();
        }
    }

    //Reset Agent and Environment
    public override void AgentReset(){
        Debug.Log("RESET "+ DateTime.Now.ToString("h:mm:ss.fff tt"));
                                                                                                                                                                                                                                                                                                                                                                               
        //Set default endEffector position and orientation
        transform.position = eeDefPosition;
        transform.rotation = eeDefRotation;
        //transform.Rotate(eeDefRotation.eulerAngles, Time.deltaTime * 800f);
        // Set new position and orientation of camera
        camera.GetComponent<CameraRandomization>().RandomPositionAndOrientation();
        // Set new position and orientation of light
        light.GetComponent<LightRandomization>().RandomPositionAndOrientation();
        // Set new light intensity
        light.GetComponent<LightRandomization>().RandomIntensity();
        // Set new dimension of boxes along x
        target.GetComponent<ScaleRandomization>().RandomizeFunction();
        target2.GetComponent<ScaleRandomization>().RandomizeFunction();
        target3.GetComponent<ScaleRandomization>().RandomizeFunction();
        // Set new position and orientation of boxes
        target.GetComponent<PositionRandomization>().GetBlockRandomPosRot(posBlocks[0]);
        target2.GetComponent<PositionRandomization>().GetBlockRandomPosRot(posBlocks[1]);
        target3.GetComponent<PositionRandomization>().GetBlockRandomPosRot(posBlocks[2]);

        //academy.AcademyReset();
    }

**Comment by alessiabertugli** [comment]
The problem was in hyperparameters, in particular in learning rate. I set a higher learning rate (1e^-3) and I changed the way to update it. I changed this line in models.py
self.learning_rate = tf.train.polynomial_decay(lr, self.global_step, max_step, 1e-10, power=1.0)
with
self.learning_rate = tf.train.exponential_decay(lr, self.global_step, 100000, 0.1, staircase=True)
using exponential_decay and updating learning rate of 0.1 every 100000 steps.
Now it seems to work and the error no longer appears.
